pkg/sql/catalog/tabledesc/constraint.go

```diff
@@ -11,215 +11,477 @@
+// interface.
+func (c uniqueWithoutIndexConstraint) ParentTableID() descpb.ID {
+	return c.desc.TableID
+}
+
+// IsValidReferencedUniqueConstraint implements the catalog.UniqueConstraint
+// interface.
+func (c uniqueWithoutIndexConstraint) IsValidReferencedUniqueConstraint(
+	fk catalog.ForeignKeyConstraint,
+) bool {
+	return !c.IsPartial() && descpb.ColumnIDs(c.desc.ColumnIDs).PermutationOf(fk.ForeignKeyDesc().ReferencedColumnIDs)
+}
+
+// NumKeyColumns implements the catalog.UniqueConstraint interface.
+func (c uniqueWithoutIndexConstraint) NumKeyColumns() int {
+	return len(c.desc.ColumnIDs)
+}
+
+// GetKeyColumnID implements the catalog.UniqueConstraint interface.
+func (c uniqueWithoutIndexConstraint) GetKeyColumnID(columnOrdinal int) descpb.ColumnID {
+	if columnOrdinal < 0 || columnOrdinal >= len(c.desc.ColumnIDs) {
+		return 0
 	}
+	return c.desc.ColumnIDs[columnOrdinal]
+}
 
-	// Glean all constraints from `desc` and `mutations`.
-	var allConstraints []catalog.Constraint
-	constraintIDs := make(map[descpb.ConstraintID]bool)
-	for _, cst := range []descpb.IndexDescriptor{desc.PrimaryIndex} {
-		backingStruct := index{
-			desc:                 &cst,
-			validityIfConstraint: descpb.ConstraintValidity_Validated,
-		}
-		allConstraints = addIfNotExists(backingStruct, allConstraints, constraintIDs)
+// CollectKeyColumnIDs implements the catalog.UniqueConstraint
+// interface.
+func (c uniqueWithoutIndexConstraint) CollectKeyColumnIDs() catalog.TableColSet {
+	return catalog.MakeTableColSet(c.desc.ColumnIDs...)
+}
+
+// IsPartial implements the catalog.UniqueConstraint interface.
+func (c uniqueWithoutIndexConstraint) IsPartial() bool {
+	return c.desc.Predicate != ""
+}
+
+// GetPredicate implements the catalog.UniqueConstraint interface.
+func (c uniqueWithoutIndexConstraint) GetPredicate() string {
+	return c.desc.Predicate
+}
+
+// GetConstraintID implements the catalog.Constraint interface.
+func (c uniqueWithoutIndexConstraint) GetConstraintID() descpb.ConstraintID {
+	return c.desc.ConstraintID
+}
+
+// GetConstraintValidity implements the catalog.Constraint interface.
+func (c uniqueWithoutIndexConstraint) GetConstraintValidity() descpb.ConstraintValidity {
+	return c.desc.Validity
+}
+
+// IsConstraintValidated implements the catalog.Constraint interface.
+func (c uniqueWithoutIndexConstraint) IsConstraintValidated() bool {
+	return c.desc.Validity == descpb.ConstraintValidity_Validated
+}
+
+// IsConstraintUnvalidated implements the catalog.Constraint interface.
+func (c uniqueWithoutIndexConstraint) IsConstraintUnvalidated() bool {
+	return c.desc.Validity == descpb.ConstraintValidity_Unvalidated
+}
+
+// GetName implements the catalog.Constraint interface.
+func (c uniqueWithoutIndexConstraint) GetName() string {
+	return c.desc.Name
+}
+
+// AsUniqueWithoutIndex implements the catalog.ConstraintProvider interface.
+func (c uniqueWithoutIndexConstraint) AsUniqueWithoutIndex() catalog.UniqueWithoutIndexConstraint {
+	return &c
```

Comment:
why `&c`, not `c`?

Reply:
In go, an interface is basically a pair of pointers, so it makes sense for the underlying type of what the interface is wrapping to also be a pointer type.